packages/doenetml-worker-rust/lib-doenetml-core/src/dep_graph/graph.rs

```diff
@@ -0,0 +1,307 @@
+        nodes: &'a [Node],
+        iter: std::vec::IntoIter<usize>,
+    }
+
+    impl<'a, Node> DescendantTopologicalIterator<'a, Node> {
+        pub fn new(nodes: &'a [Node], edges: &'a [Vec<usize>], start_index: usize) -> Self {
+            let rti = ReverseTopologicalIterator::new(edges, start_index);
+            let mut order = rti.collect::<Vec<_>>();
+            order.reverse();
+
+            Self {
+                nodes,
+                iter: order.into_iter(),
+            }
+        }
+    }
+
+    impl<'a, Node> Iterator for DescendantTopologicalIterator<'a, Node> {
+        type Item = &'a Node;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            self.iter.next().map(|index| &self.nodes[index])
+        }
+    }
+
+    /// An iterator that yields indices of all descendants of a node in a graph in reverse topological order.
+    pub struct ReverseTopologicalIterator<'a> {
+        edges: &'a [Vec<usize>],
+        remaining_indices: Vec<usize>,
+        visited: Vec<bool>,
+    }
+
+    impl<'a> ReverseTopologicalIterator<'a> {
+        pub fn new(edges: &'a [Vec<usize>], start_index: usize) -> Self {
+            let visited = vec![false; edges.len()];
+            let remaining_indices = edges[start_index].clone();
+            Self {
+                edges,
+                visited,
+                remaining_indices,
+            }
+        }
+    }
+
+    impl<'a> Iterator for ReverseTopologicalIterator<'a> {
+        type Item = usize;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            // We want to return nodes sorted topologically.
+            while let Some(&index) = self.remaining_indices.last() {
+                // In the worst case, we have to queue the whole graph for iteration.
+                // If we have more queued than that, we have a cycle.
+                if self.remaining_indices.len() > self.edges.len() {
+                    panic!("Cycle detected in graph")
+                }
+
+                // If the node has been visited, pop it off the stack and continue.
+                if self.visited[index] {
+                    self.remaining_indices.pop();
+                    continue;
+                }
+                // If the node has no unvisited children, that's the node we want
+                // to return next.
+                if self.edges[index].iter().all(|&to| self.visited[to]) {
+                    self.visited[index] = true;
+                    self.remaining_indices.pop();
+                    return Some(index);
+                }
+                // Otherwise, add all unvisited children to the stack.
+                self.remaining_indices
+                    .extend(self.edges[index].iter().filter(|&&to| !self.visited[to]));
+            }
+            None
+        }
+    }
+}
+
+/// `Taggable` items can have data attached to them. This provides hash-map-like functionality,
+/// but may be optimized for faster lookup.
```

Comment:
In our potential, optimized use case, `Node` would be `usize`, i.e., the index corresponding to a bigger object stored outside of `Graph`, and then the `Taggable` would be a vector with `Node` being the index in the vector? In this case, we'd just pad the vector if we add an index out of order (which will definitely happen when adding edges or reverse edges)?

Reply:
OK, I see how are you doing it in `prop_graph.rs`, where `Node` is a multi-index.

Reply:
Yeah. This is meant to be implemented as an efficient sparse data structure for the `(comp_id, prop_id)` pairs. It's generalized so that it can store any information. I assume it is going to be used to store chache data/etc. as well, but the `Graph` object only needs it for `usize`s.